doc/read-the-docs-site/adr/0005-inlining-saturated-functions.md

```diff
@@ -0,0 +1,206 @@
+The arity of a function is the number of arguments it takes minus the number of arguments applied.
+E.g., `\x.\y.x+y` has an arity of 2. `(\x.\y.x+y) 1` has an arity of 1.
+
+In this inline pass we check for a binding that is a function and is fully applied in the body.
+
+E.g.,
+
+```haskell
+let v = rhs in body
+```
+
+We pattern match the _rhs_ with `LamAbs` or `TyAbs` (lambda abstraction for terms or types),
+and count the number of lambdas.
+Then, in the _body_, we check for term or type application (`Apply` or `TyInst`)of _v_.
+
+If _v_ is fully applied in the body, i.e., if
+
+1. the number of type lambdas equals the number of type arguments applied, and
+2. the number of term lambdas equals the number of term arguments applied,
+
+we inline _v_, i.e., replace its occurrence with _rhs_ in the _body_ (**if other call site inlining
+conditions are satisfied**. We currently inline too little, this will be improved later, see below).
+For the rest of the discussion here we focus on the fully applied condition only.
+
+E.g., for
+
+```haskell
+let f = \x.\y.x+y in
+    f 2 4
+```
+
+the _body_ `f 2 4` is a function application of `f`. Since `f` is fully applied, we inline `f`.
+
+### Cases we miss
+
+The above case is straight forward. But there are other cases that
+may not give us the desired result if we just count the lambdas. Below are some examples.
+
+#### Example 1: function in body
+
+```haskell
+let f = \x . x
+     g = \y . f
+in g a
+```
+
+`f` and `g` each has an arity of 1. However, `g`'s _body_ includes `f` which also has an arity of 1.
+Since we only count the number of lambdas, `g` is fully applied, and we inline.
+But `g a` reduces to `f`, which has an arity of 1.
+So inlining `g` doesn't give the desired result of a fully reducible term in this case.
+
+#### Example 2: function as an argument
+
+```haskell
+let id :: forall a . a -> a
+    id x = x
+in id {int -> int} (\x -> x +1)
+```
+
+Here we have a type application for a function that takes one type variable.
+I.e., it's fully applied in terms of type.
+In terms of terms, `id` takes one argument, and is indeed applied to one.
+So `id` is fully applied! And we inline it.
+
+However, reducing the _body_ `id {int -> int} (\x -> x +1)` leaves us `(\x -> x +1)`,
+a function of arity 1, not 0.
+
+In this case, inlining and reducing `id` reduces the amount of code,
+so even though it reduces to a not fully applied function, it may be desirable.
+Again, we need more sophisticated heuristics to improve the performance of this.
+
+#### Example 3: function application in _RHS_
+
+```haskell
+let f = (\x.\y.x+y) 4
+in f 5
+```
+
+With beta-reduction, `f` becomes `\y.4+y` and it has an arity of 1.
+The _body_ `f 5` has an arity of 0 and thus is a fully applied function!
+We can reduce it to 9.
+However, because the _rhs_ `(\x.\y.x+y) 4` is a function application, not a lambda abstraction,
+we won't inline it with the current implementation.
```

Comment:
It depends on the order of optimizations. If we run beta reduction first, `f` will become `\y . 4+y`. If beta reduction runs after inlining fully saturated functions, it can still be inlined in the next simplifier pass (by default we run 4 passes).